File: linkedList.java
import java.util.Scanner;

public class LinkedList1 {

    static class Node {
        String nim, name;
        int order;
        Node next;

        Node(String nim, String name, int order) {
            this.nim = nim;
            this.name = name;
            this.order = order;
        }
    }

    static Node head = null;
    static int size = 0;
    static int counter = 0;
    static final int MAX = 10;
    static Scanner input = new Scanner(System.in);

    static void menu() {
        System.out.println("\n=== STUDENT MENU ===");
        System.out.println("1. Insert at beginning");
        System.out.println("2. Insert at given position");
        System.out.println("3. Insert at end");
        System.out.println("4. Delete from beginning");
        System.out.println("5. Delete given position");
        System.out.println("6. Delete from end");
        System.out.println("7. Delete first occurence");
        System.out.println("8. Show data");
        System.out.println("9. Exit");
    }

    static String inputNIM() {
        while (true) {
            System.out.print("ENTER NIM: ");
            String nim = input.nextLine();
            if (nim.matches("\\d+")) return nim;
            System.out.println("NIM MUST BE NUMBERS ONLY!");
        }
    }

    static String inputName() {
        while (true) {
            System.out.print("ENTER NAME: ");
            String name = input.nextLine();
            if (name.matches("[a-zA-Z ]+")) return name;
            System.out.println("NAME MUST BE LETTERS ONLY!");
        }
    }

    static Node createNode() {
        if (size == MAX) {
            System.out.println("DATA IS FULL!");
            return null;
        }
        counter++;
        return new Node(inputNIM(), inputName(), counter);
    }

    static void insertAtBeginning() {
        Node newNode = createNode();
        if (newNode == null) return;

        newNode.next = head;
        head = newNode;
        size++;
    }

    static void insertAtEnd() {
        Node newNode = createNode();
        if (newNode == null) return;

        if (head == null) {
            head = newNode;
        } else {
            Node temp = head;
            while (temp.next != null)
                temp = temp.next;
            temp.next = newNode;
        }
        size++;
    }

    static void insertAtPosition() {
        if (size == MAX) {
            System.out.println("DATA IS FULL!");
            return;
        }

        System.out.print("ENTER POSITION (1-" + (size + 1) + "): ");
        if (!input.hasNextInt()) {
            System.out.println("POSITION MUST BE NUMBER!");
            input.nextLine();
            return;
        }

        int pos = input.nextInt();
        input.nextLine();

        if (pos < 1 || pos > size + 1) {
            System.out.println("INVALID POSITION!");
            return;
        }

        if (pos == 1) {
            insertAtBeginning();
            return;
        }

        Node newNode = createNode();
        if (newNode == null) return;

        Node temp = head;
        for (int i = 1; i < pos - 1; i++)
            temp = temp.next;

        newNode.next = temp.next;
        temp.next = newNode;
        size++;
    }

    static void deleteAtBeginning() {
        if (size == 0) {
            System.out.println("EMPTY DATA!");
            return;
        }
        head = head.next;
        size--;
    }

    static void deleteAtEnd() {
        if (size == 0) {
            System.out.println("EMPTY DATA!");
            return;
        }
        if (head.next == null) {
            head = null;
        } else {
            Node temp = head;
            while (temp.next.next != null)
                temp = temp.next;
            temp.next = null;
        }
        size--;
    }

    static void deleteAtPosition() {
        if (size == 0) {
            System.out.println("EMPTY DATA!");
            return;
        }
        System.out.print("ENTER POSITION (1-" + size + "): ");
        if (!input.hasNextInt()) {
            System.out.println("POSITION MUST BE NUMBER!");
            input.nextLine();
            return;
        }

        int pos = input.nextInt();
        input.nextLine();

        if (pos < 1 || pos > size) {
            System.out.println("INVALID POSITION!");
            return;
        }
        if (pos == 1) {
            deleteAtBeginning();
            return;
        }

        Node temp = head;
        for (int i = 1; i < pos - 1; i++)
            temp = temp.next;

        temp.next = temp.next.next;
        size--;
    }

    static void deleteFirstInserted() {
        if (size == 0) {
            System.out.println("EMPTY DATA!");
            return;
        }

        Node temp = head;
        Node prev = null;
        Node target = head;
        Node targetPrev = null;

        int smallest = head.order;

        while (temp != null) {
            if (temp.order < smallest) {
                smallest = temp.order;
                target = temp;
                targetPrev = prev;
            }
            prev = temp;
            temp = temp.next;
        }

        if (targetPrev == null)
            head = target.next;
        else
            targetPrev.next = target.next;

        size--;
        System.out.println("FIRST INSERTED DATA DELETED.");
    }

    static void showData() {
        if (size == 0) {
            System.out.println("EMPTY DATA!");
            return;
        }

        Node temp = head;
        int i = 1;
        while (temp != null) {
            System.out.println(i + ". NIM: " + temp.nim + " | Name: " + temp.name);
            temp = temp.next;
            i++;
        }
    }

    public static void main(String[] args) {
        int choice;

        do {
            menu();
            System.out.print("SELECT MENU: ");

            if (input.hasNextInt()) {
                choice = input.nextInt();
                input.nextLine();

                switch (choice) {
                    case 1 -> insertAtBeginning();
                    case 2 -> insertAtPosition();
                    case 3 -> insertAtEnd();
                    case 4 -> deleteAtBeginning();
                    case 5 -> deleteAtPosition();
                    case 6 -> deleteAtEnd();
                    case 7 -> deleteFirstInserted();
                    case 8 -> showData();
                    case 9 -> System.out.println("PROGRAM COMPLETED.");
                    default -> System.out.println("INVALID MENU!");
                }
            } else {
                System.out.println("INPUT MUST BE NUMBER!");
                input.nextLine();
                choice = 0;
            }

        } while (choice != 9);
    }
}

File: linkedList.py
MAX = 10
head = None
size = 0
counter = 0

class Node:
    def __init__(self, nim, name, order):
        self.nim = nim
        self.name = name
        self.order = order
        self.next = None

def input_nim():
    while True:
        nim = input("Input NIM (numbers only): ")
        if nim.isdigit():
            return nim
        print("NIM must be numbers only!")

def input_name():
    while True:
        name = input("Input Name (letters only): ")
        if name.replace(" ", "").isalpha():
            return name
        print("Name must be letters only!")

def insert(pos):
    global head, size, counter

    if size == MAX:
        print("DATA IS FULL!")
        return

    nim = input_nim()
    name = input_name()

    counter += 1
    new_node = Node(nim, name, counter)

    if pos == 0:
        new_node.next = head
        head = new_node
    else:
        temp = head
        for _ in range(pos - 1):
            temp = temp.next
        new_node.next = temp.next
        temp.next = new_node

    size += 1
    print("Data inserted successfully.")

def delete(pos):
    global head, size

    if size == 0:
        print("EMPTY DATA!")
        return

    if pos == 0:
        head = head.next
    else:
        temp = head
        for _ in range(pos - 1):
            temp = temp.next
        temp.next = temp.next.next

    size -= 1
    print("Data deleted successfully.")

def delete_first_inserted():
    global head, size

    if size == 0:
        print("EMPTY DATA!")
        return

    temp = head
    prev = None
    smallest = temp.order
    target = temp
    target_prev = None

    while temp:
        if temp.order < smallest:
            smallest = temp.order
            target = temp
            target_prev = prev
        prev = temp
        temp = temp.next

    if target_prev is None:
        head = target.next
    else:
        target_prev.next = target.next

    size -= 1
    print("First inserted data deleted.")

def display():
    if size == 0:
        print("EMPTY DATA!")
        return

    temp = head
    print("\nCurrent Data:")
    while temp:
        print("NIM:", temp.nim, "| Name:", temp.name)
        temp = temp.next
    print()

while True:
    print("\n=== STUDENT MENU ===")
    print("1. Insert at beginning")
    print("2. Insert at given position")
    print("3. Insert at end")
    print("4. Delete from beginning")
    print("5. Delete from given position")
    print("6. Delete from end")
    print("7. Delete first occurence")
    print("8. Show data")
    print("9. Exit")

    choice = input("Choose menu: ")

    if choice == "1":
        insert(0)

    elif choice == "2":
        if size == 0:
            print("EMPTY DATA!")
        else:
            try:
                pos = int(input("Position (1 - {}): ".format(size + 1))) - 1
                if 0 <= pos <= size:
                    insert(pos)
                else:
                    print("Invalid position!")
            except:
                print("Position must be number!")

    elif choice == "3":
        insert(size)

    elif choice == "4":
        delete(0)

    elif choice == "5":
        if size == 0:
            print("EMPTY DATA!")
        else:
            try:
                pos = int(input("Position (1 - {}): ".format(size))) - 1
                if 0 <= pos < size:
                    delete(pos)
                else:
                    print("Invalid position!")
            except:
                print("Position must be number!")

    elif choice == "6":
        delete(size - 1)

    elif choice == "7":
        delete_first_inserted()

    elif choice == "8":
        display()

    elif choice == "9":
        print("Program finished.")
        break

    else:
        print("Invalid choice!")
